misc/clock.ru

```diff
@@ -71,16 +84,34 @@ loop do
   # and the data to publish.
 
   # The Redis approach.
+  # 
+  # I have a preference for this one just because that's what I did in the past,
+  # (I used 'channel' and 'pattern' as the two names)...
+  # ...but we can surely use the 'subject' name if you feel it's better. I just
+  # think it might be a bit confusing where pattern matching is concerned.
   Rack.publish(channel: :time,
 	       message: "%02d:%02d:%02d" % [now.hour, now.min, now.sec])
   # For publishing channel and subject would appear the same. Maybe even alias
   # channel and subject.
+  # 
+  # I prefer deciding on a name rather than aliasing. Creating an alias is a
+  # good tool for backwards compatibility, but isn't super clear for a new spec.
+  # 
+  # The only naming preference I ask for is that it is consistent with naming
+  # for 'subscribe'	
```

Comment:
I agree, being consistent is my preference as well. We just have to figure out which approach to take.

Reply:
I don't mind calling it `subject`.

I also like your idea a little better now that I understand what you mean by `subject`.

This allows 3 types of subscriptions: exact match, pattern and automatic (detects pattern vs. exact match).

It's kinda cool to support and only requires a simple loop over the string to detect unescaped `*`, `?` or `[`... (`+` isn't supported by Redis, as far as I understand glob matching vs. pattern matching, or am I wrong?).

On the other hand, wouldn't the automatic detection mode be backend specific? Wouldn't some support glob matching and others support pattern matching, making automatic detection a bit hazardous?

Reply:
A common set of wild cards for subjects are * and > where * matches one token and > matches the rest. It makes pattern matching fast and simple to implement. There is some precedence as well. Auto detect is easy. If it has a * or > it is a pattern matching. Of course that only makes sense on subscribe and not publish.

Reply:
> A common set of wild cards for subjects are * and > ...

Oh... I think this exposes an issue with automatic detection.

For example, [Redis follows a different glob matching flavor](https://github.com/antirez/redis/blob/398b2084af067ae4d669e0ce5a63d3bc89c639d3/src/util.c#L46-L167), something similar to the style used by some file systems (except it's binary safe).

This seems to indicate that there's a number of variations "in the wild" and perhaps the automatic mode isn't as safe to implement as we might hope.

Reply:
Let’s continue the discussing the merits of each. I think we can pick something and claim that as the spec.

Reply:
I like glob matching because it's the same way the Linux and OS X kernel match file names and because most of my implementations use Redis for the backend (which, unlike [NAT](https://nats.io/documentation/faq/#wildcards), doesn't recognize `.` as a separator)...

... but these are purely selfish reasons....

... well, except the fact that Rails (through ActiveCable and AnyCable) has better support for Redis as well, so maybe it's not just me.

On the other hand, if we allow the pub/sub API to be extended (see my comment at the top of the page), than this might be easily changed by gems and other "add-ons" that could perform their own pattern matching.

Because this could be easily changed regardless of the server in question, maybe the specification should merely recommend a style rather than enforce one.

What do you think?

Reply:
I like that the idea of being flexible in that area. Glob is nice for file systems but is more expensive to performance wise to implement for high speed filtering or switching.

Reply:
> Glob is nice for file systems but is more expensive...

Yeah, I guess you're right. I never mulled this over, since I had to conform to Redis, but the NAT approach is definitely friendlier and imposes a limit on matching concerns.

I mean, it's nice to have `user*`, but `*user` is performance intensive. The assumption that the developer knows this is beautifully avoided by NAT.

[RabbitMQ also adopted a similar solution](https://www.rabbitmq.com/tutorials/tutorial-five-ruby.html) to NAT, but they use `*` and `#` instead of `*` and `>`, while [Kafka uses regular expressions](https://kafka.apache.org/0110/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#subscribe(java.util.regex.Pattern,%20org.apache.kafka.clients.consumer.ConsumerRebalanceListener)) for pattern matching (performance be damned)... which is another inconsistency to consider.

I feel tied down by what seems to be the most commonly used DataBase in Ruby (which seems to be Radis, due to Rails and it's ease of use), otherwise I would totally adopt your approach.

For now, I think I'll just be happy to imagine that patterns aren't used so much during production so it isn't super critical. I might be wrong, but the only use-case I found for them was for administrative and debugging chores.

Reply:
Not sure about production with Rack Push but in the message based system I've built in the past (many) and even the one I am working on for work right now we use the pattern approach quite a bit. As a trivial example, for logging. Publish on `domain.log.my-app` and listen on `domain.log.>`. You can listen for all logs from any app ignoring stuff like status or performance data which would go out on different subjects. Super useful when implementing specialized stream handlers.

Reply:
I'm happy you have experience using patterns (I only did it when running automated health / debugging observers).

I wonder if we the specification should require more than a single pattern support.

i.e., have a different keyword for different patterns, `redis_pattern` match against Redis glob matching, have `nats_subject` follow NAT and maybe `rabbit` follow RabbitMQ patterns.

Maybe a cleaner option would be to support a callback that performs pattern matching, i.e.:

```ruby
# I'm throwing in another API style to the list
subscribe "log.*", match: redis, as: text
# and/or
subscribe "log.>", match: :nats

# v.s.

MATCHING_CALLBACK = Proc. new {|patter, channel | pattern.match? channel }
subscribe "log.*", match: MATCHING_CALLBACK
```

It's not too much work (though I will have to rewrite the pub/sub system to accept function pointers).

Reply:
I'd like to avoid forcing a Ruby eval for filtering. Ruby calls are expensive. How about the minimum requirement is a simple string match and anything more than that is implementation specific. That sets a baseline but leave more advances options open. I would hope that anything that uses advanced options uses configurable and not hardcoded patterns.

Reply:
> I'd like to avoid forcing a Ruby eval for filtering. Ruby calls are expensive. 

That's a very fair point.

I totally agree that it would be a very expensive to have Ruby calls for filtering...

...but I do feel that it's important to allow `subscribe` to be local as well as direct.

The most common use case for this is logging.

i.e.:

```ruby
Rack::Push.subscribe("*") {|ch, msg| log_message(ch, msg) }
```

This is a local application subscription that could be super handy.

> How about the minimum requirement is a simple string match and anything more than that is implementation specific.

I like that... but I believe that this does require that we use a Hash to allow implementations to offer extended options.

The specification could recommend that implementations support different matching algorithms using the `:match` key in the options Hash.

Since it's likely that common options will be implemented (such as pattern matching), it makes sense to have recommendations in place to guide their evolution.

i.e.

> Subscribe MUST be defined to .
>
>```ruby
>def subscribe(name, options = {}, &optional_block)
>    # implementation specific
>end
>```
>
> `options` is an implementation specific options Hash. Implementations MUST ignore `options` if it isn't a Hash (i.e., if it's `nil`).
>
> If the optional block is provided, implementations MUST call the block instead of sending the message directly i.e.:
>
>```ruby
> subscribe("char.room.1") {|channel, message| puts channel, message}
>```
>
> It is recommended that implementations support pattern matching using the `:match` options key with possible Symbol (or Proc) values, i.e.:
>
>```ruby 
>subscribe "my.subject.>", match: :nats
> # or
>subscribe "my.subject.*", match: :redis
>```
>
> It is recommended that implementations that support binary transport use the `:as` option key with the values `:binary` or `:text`. i.e.:
>
>```ruby
>subscribe "my.subject.hello", as: :binary
>```
>
> Implementations that don't support binary transport MUST default to text???? (best guess?)
>
> Binary transport over SSE, if implemented, MUST use Base64 encoding to encode the message before transmission.

Reply:
Glad you provided that code snippet. I think we are thinking along two different lines. To me a subscription is for the JavaScript associated with a connection established by one of the extended handlers. You example seems to expect published messages to be distributed to a set of Ruby objects of blocks which has nothing to do with WebSockets or SSE. While I like the pub/sub model and am currently using an in process pub/sub in some go code I think it should be separate from Rack Push. Have I missed something? Maybe we both heard what we were expecting.

Reply:
> To me a subscription is for the JavaScript associated with a connection established by one of the extended handlers.

This is the main use case I'm thinking of. This will definitely have the best performance and be the most meaningful feature.

> You example seems to expect published messages to be distributed to a set of Ruby objects of blocks which has nothing to do with WebSockets or SSE.

This is more of a side effect, or a necessary evil.

I see the possible use cases for server-side pub/sub that isn't being forwarded to a connection. Logging is such a use case, as well as message filtering or re-writing.

I believe that support for these added use cases is unavoidable.